docs/question/server.md

```diff
@@ -1,89 +1,321 @@
+| `file()`     | :x:                | N/A. Returns an `object` (string, bytes-like, file-like)                                                 | Generate a file object dynamically in lieu of a physical file. Trigger via `type="dynamic"` in the question element (e.g., `pl-figure`, `pl-file-download`). Access the requested filename via `data['filename']`. If `file()` returns nothing, an empty string will be used. |
 
-To account for this, the `prairielearn` Python library (usually aliased and used as `pl`), provides the functions [`to_json`][prairielearn.conversion_utils.to_json] and [`from_json`][prairielearn.conversion_utils.from_json] (part of [`conversion_utils.py`](https://github.com/PrairieLearn/PrairieLearn/blob/master/apps/prairielearn/python/prairielearn/conversion_utils.py)), which can respectively serialize and deserialize various objects for storage as part of question data. Please refer to the docstrings on those functions for more information. Here is a simple example:
+As shown in the table, most functions accept a single argument, `data` (a dictionary), and modify it in place.
+
+### `data` dictionary
+
+| Attribute               | Type    | Description                                                                                                                          |
+| ----------------------- | ------- | ------------------------------------------------------------------------------------------------------------------------------------ |
+| `params`                | `dict`  | Parameters for the question variant. These are set in the `generate()` function and can be used in the `question.html` file.         |
+| `correct_answers`       | `dict`  | Correct answers for the question variant. Each item maps from a named answer to a value.                                             |
+| `submitted_answers`     | `dict`  | Student answers submitted for the question after parsing.                                                                            |
+| `raw_submitted_answers` | `dict`  | Raw student answers submitted for the question.                                                                                      |
+| `format_errors`         | `dict`  | Dictionary of format errors for each answer. Each item maps from a named answer to a error message.                                  |
+| `partial_scores`        | `dict`  | Dictionary of partial scores for each answer. Each entry is a dictionary with the keys `score` (float) and `weight` (int, optional). |
+| `score`                 | `float` | The total score for the question variant.                                                                                            |
+| `feedback`              | `dict`  | Dictionary of feedback for each answer. Each item maps from a named answer to a feedback message.                                    |
```

Comment:
Missing: `variant_seed` (all phases), `options` (all phases), `filename` (for `file()`). There are also a few specific to `render()` (`editable`, `manual_grading`, `panel`, `num_valid_submissions`), `prepare()` (`answers_name`) and `test()` (`test_type`), but I don't think they need to be listed since the docs below point to them.

Reply:
To be clear: I don't think they need extensive documentation, but at least a link to where they are described may be useful. All three of them have a description in the same file.

Reply:
Changed in 6408ede3ac9fe51ab8d193ea476887f9d181752a.